share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java

```diff
@@ -0,0 +1,519 @@
+        // onto the shard method.
+
+        request.topics().forEach(topicData -> {
+            Map<Integer, CompletableFuture<WriteShareGroupStateResponseData>> partitionFut =
+                futureMap.computeIfAbsent(topicData.topicId(), k -> new HashMap<>());
+            topicData.partitions().forEach(
+                partitionData -> partitionFut.put(partitionData.partition(), runtime.scheduleWriteOperation(
+                    "write-share-group-state",
+                    topicPartitionFor(SharePartitionKey.getInstance(groupId, topicData.topicId(), partitionData.partition())),
+                    Duration.ofMillis(config.shareCoordinatorOffsetsCommitTimeoutMs()),
+                    coordinator -> coordinator.writeState(context, new WriteShareGroupStateRequestData()
+                        .setGroupId(groupId)
+                        .setTopics(Collections.singletonList(new WriteShareGroupStateRequestData.WriteStateData()
+                            .setTopicId(topicData.topicId())
+                            .setPartitions(Collections.singletonList(new WriteShareGroupStateRequestData.PartitionData()
+                                .setPartition(partitionData.partition())
+                                .setStartOffset(partitionData.startOffset())
+                                .setLeaderEpoch(partitionData.leaderEpoch())
+                                .setStateEpoch(partitionData.stateEpoch())
+                                .setStateBatches(partitionData.stateBatches()))))))))
+            );
+        });
+
+        // Combine all futures into a single CompletableFuture<Void>
+        CompletableFuture<Void> combinedFuture = CompletableFuture.allOf(futureMap.values().stream()
+            .flatMap(partMap -> partMap.values().stream()).toArray(CompletableFuture[]::new));
+
+        return combinedFuture.thenApply(v -> {
+            List<WriteShareGroupStateResponseData.WriteStateResult> writeStateResults = futureMap.keySet().stream()
+                .map(topicId -> {
+                    List<WriteShareGroupStateResponseData.PartitionResult> partitionResults = futureMap.get(topicId).entrySet().stream()
+                        .map(topicEntry -> {
+                            CompletableFuture<WriteShareGroupStateResponseData> future = topicEntry.getValue();
+                            int partition = topicEntry.getKey();
+                            try {
+                                long timeTaken = time.hiResClockMs() - startTime;
+                                WriteShareGroupStateResponseData partitionData = future.get();
+                                // This is the future returned by runtime.scheduleWriteOperation which returns when the
+                                // operation has completed including
+                                shareCoordinatorMetrics.globalSensors.get(ShareCoordinatorMetrics.SHARE_COORDINATOR_WRITE_LATENCY_SENSOR_NAME)
```

Comment:
I'm not sure this is the pattern we want for metrics. In GC, I see we are calling `record(String sensorName)` rather than accessing the globalSensors